kernel/src/net/socket/ip/stream/mod.rs

```diff
@@ -651,11 +670,23 @@ impl Socket for StreamSocket {
     }
 
     fn set_option(&self, option: &dyn SocketOption) -> Result<()> {
-        let mut options = self.options.write();
+        let (mut options, mut state) = self.update_connecting();
 
-        match options.socket.set_option(option) {
+        match options.socket.set_option(option, state.as_mut()) {
             Err(err) if err.error() == Errno::ENOPROTOOPT => (),
-            res => return res,
+            Ok(true) => {
+                let iface_to_poll = state.iface_to_poll();
+
+                drop(state);
+                drop(options);
+
+                if let Some(iface) = iface_to_poll {
+                    iface.poll();
+                }
+
+                return Ok(());
+            }
+            res => return res.map(|_| ()),
```

Comment:
This can be made to be more consistent with how `try_send`/`try_recv` deals with the conditional polling.
 ```suggestion
            Err(err) => return Err(err),
            Ok(need_poll) => {
                let iface_to_poll = need_poll.then(|| state.iface().cloned()).flatten();

                drop(state);
                drop(options);
                if let Some(iface) = iface_to_poll {
                    iface.poll();
                }

                return Ok(());
            }
```